ground/src/main/java/com/google/android/ground/persistence/sync/MediaUploadWorker.kt

```diff
@@ -0,0 +1,184 @@
+    //   reattempted if they satisfy the retry count bound. Eventually, either all of the
+    //   mutations for this LOI will have succeeded, failed, or exceeded the retry count.
+    if (
+      mutationStatusesAfterUpload.any {
+        it == Mutation.SyncStatus.FAILED || it == Mutation.SyncStatus.MEDIA_UPLOAD_PENDING
+      }
+    ) {
+      return Result.retry()
+    }
+    return Result.success()
+  }
+
+  /**
+   * Attempts to upload associated media for a given list of [SubmissionMutation] and updates the
+   * status of each mutation based on whether uploads succeeded or failed.
+   */
+  private suspend fun uploadMediaAndGetStatus(
+    mutations: List<SubmissionMutation>
+  ): List<Mutation.SyncStatus> =
+    mutations
+      .map { uploadSubmissionMediaAndUpdateMutation(it) }
+      .also { resultingMutations -> mutationRepository.updateMutations(resultingMutations) }
+      .map { it.syncStatus }
+
+  /**
+   * Attempts to upload all media associated with a given submission. Updates the submission's sync
+   * status depending on whether or the uploads failed or succeeded.
+   */
+  private suspend fun uploadSubmissionMediaAndUpdateMutation(
+    mutation: SubmissionMutation
+  ): SubmissionMutation =
+    uploadSubmissionPhotos(mutation)
+      .fold(
+        onSuccess = { mutation.copy(syncStatus = Mutation.SyncStatus.COMPLETED) },
+        onFailure = {
+          val cause = it.message ?: "unknown upload error"
+          if (it is FileNotFoundException) {
+            return mutation.copy(syncStatus = Mutation.SyncStatus.FAILED, lastError = cause)
+          }
+
+          return mutation.copy(
+            syncStatus = Mutation.SyncStatus.MEDIA_UPLOAD_PENDING,
+            lastError = cause
+          )
+        }
+      )
+
+  private suspend fun uploadSubmissionPhotos(mutation: SubmissionMutation): kotlin.Result<Unit> =
+    // TODO: Use media response types instead of discriminating on Task.Type.
+    // For example, we should pass a List<PhotoResponse> to uploadPhotoMedia(), which can take care
+    // of the bulk of the response-specific work.
+    mutation.deltas
+      .filter { (_, taskType, newValue): ValueDelta ->
+        taskType === Task.Type.PHOTO && newValue.isNotNullOrEmpty()
+      }
+      .map { (_, _, newValue): ValueDelta -> newValue.toString() }
+      .map { uploadPhotoMedia(it) }
+      .reduce { a, b -> combineResults(a, b) }
+
+  /**
+   * Attempts to upload a single photo to remote storage. Returns an [UploadResult] indicating
+   * whether the upload attempt failed or succeeded.
+   */
+  private suspend fun uploadPhotoMedia(path: String): kotlin.Result<Unit> {
+    val photoFile = userMediaRepository.getLocalFileFromRemotePath(path)
+    return if (!photoFile.exists()) {
+      Timber.e("Photo not found. local path: ${photoFile.path}, remote path: $path")
+      kotlin.Result.failure(
+        FileNotFoundException("Photo $path not found on device: ${photoFile.path}")
+      )
+    } else {
```

Comment:
Consider using short-circuit return (`if (!exists) { return ... }`) to reduce the amount of nesting here.

Reply:
I think I had that form at first but then the static analysis yelled at me for having more than 2 return statements.

Reply:
I was able to work around it thanks to Kotlin's "everything is an expression rule"